### ISSUE TITLE
Use of links between containers should be deprecated

### ISSUE THREAD

**Issue by rnorth** [testcontainers/testcontainers-java#465]
As discussed in #463, docker networks seem mature enough to use, and links are officially not recommended any more.

We'd like to deprecate use of links; off the top of my head I believe we ought to do the following things:

- [x] Obviously, mark the container `addLink` method as deprecated with a clear comment
- [ ] Links aren't currently documented very well, but update the docs with a clear enough description of the networking support, so that ad-hoc linking can be done easily by users
- [x] Modify Webdriver VNC recorder sidecar container to remove use of links
- [x] Modify Docker Compose support; can we remove use of the ambassador container now?

**Comment by kiview** [comment]
I'm really looking forward to removing the ambassador container, I believe this will make things a lot easier in more complex CI environments.

**Comment by kiview** [comment]
AFAIK the only thing missing now is the documentation update?

**Comment by felixbarny** [comment]
I'd very much appreciate some docs around networks :)

**Comment by kiview** [comment]
@felixbarny Yep you are right, PR appreciated :smile_cat:

**Comment by bsideup** [comment]
@felixbarny yes, the docs are missing and I feel sorry about it (since I was the one who implemented the networks support)
However, it is as simple as `.withNetwork(networkInstance)` and until it is implemented you an check existing usages in tests, i.e. Kafka:
https://github.com/testcontainers/testcontainers-java/blob/b5181f784c63b91a61ec68cd58acc522ce4d2dbc/modules/kafka/src/test/java/org/testcontainers/containers/KafkaContainerTest.java#L52

**Comment by rnorth** [comment]
Reopening - docs still needed

**Comment by bsideup** [comment]
@rnorth given https://www.testcontainers.org/features/networking/#advanced-networking, should we close the issue now? :)

**Comment by sunmeplz** [comment]
Hi there,
Sometimes it is not possible to add network to container as it managed with third party libs, so add-link is the single solution

**Comment by sunmeplz** [comment]
Hi @eddumelendez !
My case: I'm using quarkus framework which provide dev services based on test containers(https://github.com/quarkusio/quarkus/blob/main/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/DevServicesKafkaProcessor.java). When you run application in dev or test mode, quarkus run the container. My goal is to create additional service which will use existing container(kafka and db in my case). So I can't access testcontainer instance, just can find already running container using plain docker api

**Comment by jfrantzius** [comment]
Hi, 
the task "Modify Docker Compose support; can we remove use of the ambassador container now?" that is listed as completed in the description is unfortunately blocking https://github.com/testcontainers/testcontainers-java/issues/5351 , please see comment https://github.com/testcontainers/testcontainers-java/issues/5351#issuecomment-1838847583 there for details.

Are there any plans to replace the deprecated call to `SocatContainer.addLink()` in `ComposeDelegate.withExposedService(String, int, WaitStrategy)` with network API calls, as Testcontainer's own Javadoc suggests? https://javadoc.io/doc/org.testcontainers/testcontainers/latest/index.html :

> @Deprecated
public void addLink([LinkableContainer](https://javadoc.io/static/org.testcontainers/testcontainers/1.19.3/org/testcontainers/containers/traits/LinkableContainer.html) otherContainer, java.lang.String alias)
Deprecated. Links are deprecated (see [#465](https://github.com/testcontainers/testcontainers-java/issues/465)). Please use [Network](https://javadoc.io/static/org.testcontainers/testcontainers/1.19.3/org/testcontainers/containers/Network.html) features instead.
Description copied from interface: [Container](https://javadoc.io/static/org.testcontainers/testcontainers/1.19.3/org/testcontainers/containers/Container.html#addLink-org.testcontainers.containers.traits.LinkableContainer-java.lang.String-)
Add a link to another container.

**Comment by kiview** [comment]
Thanks for collecting the references @jfrantzius. I am not sure why the compose task in the original post in this issue is marked as done, since it is not yet done. I think it makes much sense to re-visist this, ideally with a non-breaking change that provides better compatibility within the ecosystem.

That being said, there aren't concrete plans to tackle this yet (although AFAIK @eddumelendez did a spike on this in the past).

**Comment by dzoech** [comment]
Hey @kiview, are there any plans yet about a switch to use networks?
In it's current state, compose is not usable with Podman (afaiu at least).

**Comment by GwFreak01** [comment]
@kiview +1 to this. Would like to be able to use TestContainers on OrbStack
